### PR TITLE
Upgrade tokens_lookup() for #2324

### PR DESCRIPTION
**Pull request by koheiw**
See #2324.

### CODE REVIEW

**Comment by koheiw** [comment]
I don't think I can make faster easily.
````
> library(quanteda)
> corp_guad <- readRDS("C:/Users/watan/Dropbox/Public/data_corpus_guardian2016-10k.rds")
> toks_guad <- tokens(corp_guad)
> dict2 <- newsmap::data_dictionary_newsmap_en
> 
> microbenchmark::microbenchmark(
+     lookup0 = tokens_lookup(toks_guad, dict2, exclusive = TRUE),
+     lookup1 = tokens_lookup(toks_guad, dict2, exclusive = FALSE),
+     lookup2 = tokens_lookup(toks_guad, dict2, exclusive = FALSE, append = TRUE),
+     times = 10
+ )
Unit: milliseconds
    expr      min       lq      mean    median        uq       max neval
 lookup0  871.273  886.365  905.7185  899.5215  912.0911  977.6024    10
 lookup1 1011.055 1025.667 1039.5367 1032.9771 1057.2817 1079.1447    10
 lookup2 1254.369 1289.318 1317.7168 1310.5505 1344.7586 1403.0677    10
```

**Comment by koheiw** [comment]
We could move `separator` to `quanteda_options()`.

**Comment by kbenoit** [comment]
I think this is the neatest solution. I can add tests and a bit more documentation.

We have a few oddball conditions wherein a token could match multiple keys; with different `exclusive` settings; and with the nested options set differently. I want to explore these in detail in tests.

See my comments on the issue [here](https://github.com/quanteda/quanteda/issues/2324#issuecomment-1855573894) but in short, I prefer these as arguments:
* `append_key` logical; if `TRUE`, append the dictionary key matched by the token to the token itself. 
* `concatenator` character added to separate additional information appended to the token, such as the dictionary key when `append_key = TRUE`.

**Comment by koheiw** [comment]
I am fine with `append_key` but we might want to change `capkeys` to `cap_key` then.

There are two types of characters: "_" is the concatenator and "/" is the separator in "United_States/COUNTRY", for example. The cancatenator should be taken from the object (see #2327) but the separator should be customizable via `separator` argument or options.

**Comment by koheiw** [comment]
It works in the same way as the non-exclusive mode, so two compounded words with keys will be inserted when the same patters are in different keys.
``` r
require(quanteda)
#> Loading required package: quanteda
#> Package version: 4.0.0
#> Unicode version: 15.1
#> ICU version: 74.1
#> Parallel computing: 16 of 16 threads used.
#> See https://quanteda.io for tutorials and examples.
txt <- c(d1 = "The United States is bordered by the Atlantic Ocean and the Pacific Ocean.",
         d2 = "The Supreme Court of the United States is seldom in a united state.")
toks <- tokens(txt, remove_punct = TRUE)

dict <- dictionary(list(Countries = c("* States", "Federal Republic of *"),
                        Oceans = c("* Ocean", "* States")), tolower = FALSE)

tokens_lookup(toks, dict, exclusive = FALSE, append = TRUE, separator = "/")
#> Tokens consisting of 2 documents.
#> d1 :
#>  [1] "The"                     "United_States/Countries"
#>  [3] "United_States/Oceans"    "is"                     
#>  [5] "bordered"                "by"                     
#>  [7] "the"                     "Atlantic_Ocean/Oceans"  
#>  [9] "and"                     "the"                    
#> [11] "Pacific_Ocean/Oceans"   
#> 
#> d2 :
#>  [1] "The"                     "Supreme"                
#>  [3] "Court"                   "of"                     
#>  [5] "the"                     "United_States/Countries"
#>  [7] "United_States/Oceans"    "is"                     
#>  [9] "seldom"                  "in"                     
#> [11] "a"                       "united"                 
#> [ ... and 1 more ]
```

<sup>Created on 2023-12-14 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by kbenoit** [comment]
> I am fine with `append_key` but we might want to change `capkeys` to `cap_key` then.
> 
> There are two types of characters: "_" is the concatenator and "/" is the separator in "United_States/COUNTRY", for example. The cancatenator should be taken from the object (see #2327) but the separator should be customizable via `separator` argument or options.

OK the distinction between `concatenator` and `separator` makes sense. It would make sense then to change the argument to `separator` also in `as.tokens.spacyr_parsed()`. The other PR that records the concatenator and allows this to be changed gives further flexibility that also builds on this distinction.

Let's keep `capkeys` just to not break earlier code.

**Comment by kbenoit** [comment]
On `exclusive = TRUE, append_key = TRUE`, why would we not want this behaviour?

> tokens_lookup(toks, dict, exclusive = TRUE, append_key = TRUE)
Tokens consisting of 2 documents.
d1 :
[1] "United_States/COUNTRIES" "Atlantic_Ocean/OCEANS" 
[3] "Pacific_Ocean/OCEANS"
...
```

File: R/tokens_lookup.R [hunk 1]
```diff
@@ -153,6 +153,8 @@ tokens_lookup.tokens_xptr <- function(x, dictionary, levels = 1:5,
     if (capkeys)
         key <- stri_trans_toupper(key)
     if (exclusive) {
+        if (append_key)
```
Comment (kbenoit) [inline R/tokens_lookup.R:156]: Should it, though?

**Comment by koheiw** [comment]
It might be possible to append keys in the exclusive mode too... I will look into it.

**Comment by koheiw** [comment]
``` r
library(quanteda)
#> Package version: 4.0.0
#> Unicode version: 15.1
#> ICU version: 74.1
#> Parallel computing: 16 of 16 threads used.
#> See https://quanteda.io for tutorials and examples.
corp_guad <- readRDS("C:/Users/watan/Dropbox/Public/data_corpus_guardian2016-10k.rds")
toks_guad <- tokens(corp_guad)
dict2 <- newsmap::data_dictionary_newsmap_en

microbenchmark::microbenchmark(
    lookup0 = tokens_lookup(toks_guad, dict2, exclusive = TRUE),
    lookup1 = tokens_lookup(toks_guad, dict2, exclusive = TRUE, append = TRUE),
    lookup2 = tokens_lookup(toks_guad, dict2, exclusive = FALSE),
    lookup3 = tokens_lookup(toks_guad, dict2, exclusive = FALSE, append = TRUE),
times = 10)
#> Unit: milliseconds
#>     expr      min        lq      mean    median        uq       max neval
#>  lookup0  876.112  887.5360  911.8614  914.5766  930.1005  965.0451    10
#>  lookup1 1099.526 1124.7093 1158.9215 1136.5502 1159.7642 1342.3178    10
#>  lookup2  987.304  989.5483 1009.0808 1002.0211 1023.8005 1068.7187    10
#>  lookup3 1273.866 1279.1408 1328.5255 1297.7294 1348.7943 1451.3470    10
```

<sup>Created on 2023-12-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by koheiw** [comment]
``` r
require(quanteda)
#> Loading required package: quanteda
#> Package version: 4.0.0
#> Unicode version: 15.1
#> ICU version: 74.1
#> Parallel computing: 16 of 16 threads used.
#> See https://quanteda.io for tutorials and examples.
txt <- c(d1 = "The United States is bordered by the Atlantic Ocean and the Pacific Ocean.",
         d2 = "The Supreme Court of the United States is seldom in a united state.")
toks <- tokens(txt, remove_punct = TRUE)

dict <- dictionary(list(Countries = c("* States", "Federal Republic of *"),
                        Oceans = c("* Ocean")), tolower = FALSE)


tokens_lookup(toks, dict, exclusive = TRUE, append = TRUE, separator = "/", capkeys = TRUE)
#> Tokens consisting of 2 documents.
#> d1 :
#> [1] "United_States/COUNTRIES" "Atlantic_Ocean/OCEANS"  
#> [3] "Pacific_Ocean/OCEANS"   
#> 
#> d2 :
#> [1] "United_States/COUNTRIES"
tokens_lookup(toks, dict, exclusive = FALSE, append = TRUE, separator = "/")
#> Tokens consisting of 2 documents.
#> d1 :
#>  [1] "The"                     "United_States/COUNTRIES"
#>  [3] "is"                      "bordered"               
#>  [5] "by"                      "the"                    
#>  [7] "Atlantic_Ocean/OCEANS"   "and"                    
#>  [9] "the"                     "Pacific_Ocean/OCEANS"   
#> 
#> d2 :
#>  [1] "The"                     "Supreme"                
#>  [3] "Court"                   "of"                     
#>  [5] "the"                     "United_States/COUNTRIES"
#>  [7] "is"                      "seldom"                 
#>  [9] "in"                      "a"                      
#> [11] "united"                  "state"
```

<sup>Created on 2023-12-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Comment by koheiw** [comment]
``` r
require(quanteda)
#> Loading required package: quanteda
#> Package version: 4.0.0
#> Unicode version: 15.1
#> ICU version: 74.1
#> Parallel computing: 16 of 16 threads used.
#> See https://quanteda.io for tutorials and examples.
corp_guad <- readRDS("C:/Users/watan/Dropbox/Public/data_corpus_guardian2016-10k.rds")
toks_guad <- tokens(corp_guad)
tokens_lookup(toks_guad, newsmap::data_dictionary_newsmap_en, 
              exclusive = TRUE, append_key = TRUE)
#> Tokens consisting of 10,000 documents and 8 docvars.
#> text120761 :
#> [1] "Russian/EUROPE.EAST.RU"    "Russian/EUROPE.EAST.RU"   
#> [3] "British/EUROPE.NORTH.GB"   "British/EUROPE.NORTH.GB"  
#> [5] "Russia/EUROPE.EAST.RU"     "Kazakhstan/ASIA.CENTER.KZ"
#> [7] "Kazakhstan/ASIA.CENTER.KZ"
#> 
#> text174574 :
#> character(0)
#> 
#> text141269 :
#>  [1] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#>  [3] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#>  [5] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#>  [7] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#>  [9] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#> [11] "New_York/AMERICA.NORTH.US" "New_York/AMERICA.NORTH.US"
#> [ ... and 7 more ]
#> 
#> text151432 :
#> [1] "Paris/EUROPE.WEST.FR"    "Brussels/EUROPE.WEST.BE"
#> [3] "Egypt/AFRICA.NORTH.EG"   "Russian/EUROPE.EAST.RU" 
#> 
#> text169265 :
#> [1] "US/AMERICA.NORTH.US" "UK/EUROPE.NORTH.GB" 
#> 
#> text134827 :
#>  [1] "Jordan/ASIA.WEST.JO"        "Americans/AMERICA.NORTH.US"
#>  [3] "American/AMERICA.NORTH.US"  "American/AMERICA.NORTH.US" 
#>  [5] "New_York/AMERICA.NORTH.US"  "us/AMERICA.NORTH.US"       
#>  [7] "New_York/AMERICA.NORTH.US"  "US/AMERICA.NORTH.US"       
#>  [9] "Iraq/ASIA.WEST.IQ"          "US/AMERICA.NORTH.US"       
#> [11] "American/AMERICA.NORTH.US"  "Libya/AFRICA.NORTH.LY"     
#> [ ... and 6 more ]
#> 
#> [ reached max_ndoc ... 9,994 more documents ]
```

<sup>Created on 2023-12-15 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Review by kbenoit — APPROVED** [review]

**Pull request merged**